### PR TITLE
feat: add `allowed_for_roles` decorator in `User` and `Workspace`

### LINKED ISSUE #3367: [FEATURE] Add Python decorator to check `role`
**Is your feature request related to a problem? Please describe.**

All the functions/methods related to Argilla are exposed equally to any user no matter what role they have, and we need some sort of control to avoid sending requests to Argilla if we know that an user with that role has no access in advance. Kudos to @nataliaElv for mentioning the issue that led to the proposal of this idea.

**Describe the solution you'd like**

Ideally, we should have a Python decorator that checks for the current user's role (`admin`, `owner` or `annotator` as of Argilla 1.13.0), and raise an exception if an user is trying to perform an operation that it's not allowed to via the Python client, e.g. an annotator wants to list all the users in Argilla.

**Additional context**

This feature is also nice to internally know which users have access to what, as otherwise the Python client ends up being a little bit messy and we lose control of the permission levels.

### PR DESCRIPTION
**Pull request by alvarobartt**
# Description

This PR includes the `allowed_for_roles` Python decorator that checks the role for the current user and checks whether that role is allowed to use the decorated function/method, otherwise, a `ValueError` is raised stating that the function/method cannot be used with that role and mentioning the permitted roles.

Kudos to @nataliaElv for mentioning the issue that lead to the definition of this PR.

Closes #3367

**Type of change**

- [X] New feature (non-breaking change which adds functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

- [X] Add unit tests with different unsupported roles to show that while `owner` can do everything, `admin` and `annotator` are limited

**Checklist**

- [X] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [X] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [X] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

### CODE REVIEW

File: src/argilla/client/utils.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    from typing_extensions import ParamSpec
+
+from argilla.client.api import ArgillaSingleton
+
+_P = ParamSpec("_P")
+_R = TypeVar("_R")
+
+
+def allowed_for_roles(
+    roles: List[Literal["owner", "admin", "annotator"]]
```
Comment (gabrielmbmb) [inline src/argilla/client/utils.py]: this type hint could use `UserRole` enum instead, right?

File: src/argilla/client/utils.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        >>> def delete_user(user_id: UUID) -> None:
+        >>>     ...
+    """
+
+    def decorator(func: Callable[_P, _R]) -> Callable[_P, _R]:
+        @functools.wraps(func)
+        def wrapper(*args: _P.args, **kwargs: _P.kwargs) -> _R:
+            role = ArgillaSingleton.get().user.role
+            if role not in roles:
+                raise ValueError(
```
Comment (gabrielmbmb) [inline src/argilla/client/utils.py]: maybe we can raise built-in `PermissionError` (although it's usually used with file-system permission errors) or create a custom exception

**Review by gabrielmbmb — APPROVED** [review]
LGTM! Just some comment related to the `allowed_for_roles` decorator

File: src/argilla/client/utils.py (hunk 1 above)
Comment (frascuchon) [reply src/argilla/client/utils.py]: Yes, should be better using the enum definition

**Review by frascuchon — APPROVED** [review]

**Pull request merged**